lib/src/io.dart

```diff
@@ -788,102 +791,87 @@ Future<HttpServer> bindServer(String host, int port) async {
-String _findTarPath() {
-  for (final file in ['/bin/tar', '/usr/bin/tar']) {
-    if (fileExists(file)) {
-      return file;
-    }
-  }
-  log.warning(
-      'Could not find a system `tar` installed in /bin/tar or /usr/bin/tar, '
-      'attempting to use tar from PATH');
-  return 'tar';
-}
-
 /// Extracts a `.tar.gz` file from [stream] to [destination].
 Future extractTarGz(Stream<List<int>> stream, String destination) async {
   log.fine('Extracting .tar.gz stream to $destination.');
-  final decompressed = stream.transform(GZipCodec().decoder);
 
-  // We used to stream directly to `tar`,  but that was fragile in certain
-  // settings.
-  final processResult = await withTempDir((tempDir) async {
-    final tarFile = path.join(tempDir, 'archive.tar');
-    try {
-      await _createFileFromStream(decompressed, tarFile);
-    } catch (e) {
-      // We don't know the error type here: https://dartbug.com/41270
-      throw FileSystemException('Could not decompress gz stream $e');
+  destination = path.absolute(destination);
+  final reader = TarReader(stream.transform(gzip.decoder));
+  while (await reader.moveNext()) {
+    final entry = reader.current;
+
+    final filePath = path.joinAll([
+      destination,
+      // Tar file names always use forward slashes
+      ...path.posix.split(entry.name),
+    ]);
+
+    if (!path.isWithin(destination, filePath)) {
+      // The tar contains entries that would be written outside of the
+      // destination. That doesn't happen by accident, assume that the tar file
+      // is malicious.
+      await reader.cancel();
+      throw FormatException('Invalid tar entry: ${entry.name}');
+    }
+
+    final parentDirectory = path.dirname(filePath);
+
+    switch (entry.type) {
+      case TypeFlag.dir:
+        ensureDir(filePath);
+        break;
+      case TypeFlag.reg:
+      case TypeFlag.regA:
+        // Regular file
+        deleteIfLink(filePath);
+        ensureDir(parentDirectory);
+
+        await _createFileFromStream(entry.contents, filePath);
+
+        if (Platform.isLinux || Platform.isMacOS) {
+          // Apply executable bits from tar header, but don't change r/w bits
+          // from the default
+          const defaultMode = 420; // 644 in base 8
+          const executableMask = 0x49; // 001 001 001
+          final mode = defaultMode | (entry.header.mode & executableMask);
+
+          if (mode != defaultMode) {
+            _chmod(mode, filePath);
+          }
+        }
+        break;
+      case TypeFlag.symlink:
+        // Link to another file in this tar, relative from this entry.
+        final resolvedTarget = path.joinAll(
+            [parentDirectory, ...path.posix.split(entry.header.linkName)]);
+        if (!path.isWithin(destination, resolvedTarget)) {
+          // Don't allow links to files outside of this tar.
+          break;
+        }
+
+        ensureDir(parentDirectory);
+        createSymlink(resolvedTarget, filePath);
```

Comment:
I think we want the symlink targets to be relative. Something like:
```suggestion
        createSymlink(path.relative(resolvedTarget, from: destination), filePath);
```